src/aspire/utils/blk_diag_matrix.py

```diff
@@ -0,0 +1,936 @@
+        otherwise return a new instance (default).
+        :return: The norm of the BlkDiagMatrix instance.
+        """
+
+        implemented_orders = (2,)
+        if order not in implemented_orders:
+            raise NotImplementedError(
+                "Order {} not yet implemented, only {}".format(
+                    ord, implemented_orders))
+
+        return np.max([norm(blk, ord=order) for blk in self])
+
+    def transpose(self):
+        """
+        Get the transpose matrix of a BlkDiagMatrix instance.
+
+        :return: The corresponding transpose form as a BlkDiagMatrix.
+        """
+
+        T = BlkDiagMatrix(self.partition, dtype=self.dtype)
+
+        for i in range(self.nblocks):
+            T[i] = self[i].T
+
+        return T
+
+    @property
+    def T(self):
+        """
+        Syntactic sugar for self.transpose().
+        """
+
+        return self.transpose()
+
+    def dense(self):
+        """
+        Convert list representation of BlkDiagMatrix instance into full matrix.
+
+        :param blk_diag: The BlkDiagMatrix instance.
+        :return: The BlkDiagMatrix instance including the zero elements of
+        non-diagonal blocks.
+        """
+
+        return block_diag(self.data)
+
+    @staticmethod
+    def from_list(blk_diag, dtype=np.float64):
+        """
+        Convert full from python list representation into BlkDiagMatrix.
+
+        This is to facilitate integration with code that may not be
+        using the BlkDiagMatrix class yet.
+
+        :param blk_diag; The blk_diag representation in the form of a
+        K-element list storing all shapes of K diagonal matrix blocks,
+        where `blk_partition[i]` corresponds to the shape (number of rows
+        and columns) of the `i` diagonal matrix block.
+
+        :return: The BlkDiagMatrix instance.
+        """
+
+        # get the partition (just sizes)
+        blk_partition = [None] * len(blk_diag)
+        for i, mat in enumerate(blk_diag):
+            blk_partition[i] = np.shape(mat)
+
+        # instantiate an empty BlkDiagMatrix with that structure
+        A = BlkDiagMatrix(blk_partition, dtype=dtype)
+
+        # set the data
+        for i in range(A.nblocks):
+            A.data[i] = np.array(blk_diag[i], dtype=dtype)
+
+        return A
+
+    @staticmethod
+    def from_mat(mat, blk_partition, dtype=np.float64):
+        """
+        Convert full block diagonal matrix into BlkDiagMatrix.
+
+        :param mat; The full block diagonal matrix including the zero elements
+        of non-diagonal blocks.
+        :param blk_partition: The matrix block partition in the form of a
+        K-element list storing all shapes of K diagonal matrix blocks,
+        where `blk_partition[i]` corresponds to the shape (number rows
+        and columns) of the `i` diagonal matrix block.
+
+        :return: The BlkDiagMatrix instance.
+        """
+
+        A = BlkDiagMatrix(blk_partition, dtype=dtype)
+
+        rows = blk_partition[:, 0]
```

Comment:
This doesn't seem to correspond with the standard form of a partition (list of tuples).

Reply:
Right, i used numpy syntax, thanks for catching

Reply:
Removed, per discussion.